testdata/p4_16_samples_outputs/issue2583_simplify_slice-frontend.p4

```diff
@@ -0,0 +1,20 @@
+#include <core.p4>
+
+header Header {
+    bit<32> data;
+}
+
+parser p(packet_in pckt, out Header h) {
+    state start {
+        h.data = 32w0;
+        h.data = 32w7;
```

Comment:
shouldn't this be simplified as well?

Reply:
Yes, you're right, it should be. But this makes for a great segue into another point I wanted to make. Consider this simplification (snippet):
```
parser p(packet_in pckt, out Header h, out Header h2) {
    state start {
        h.data = 8;  // A
        h2.data = h.data; // B
        h.data = 8; // C
        h2.data = 5; // D
        transition accept;
    }
}

// turns into

parser p(packet_in pckt, out Header h, out Header h2) {
    state start {
        h.data = 32w8;
        ;
        h.data = 32w8;
        h2.data = 32w5;
        transition accept;
    }
}
```
Looks familiar. Pardon the nonsensical piece of code (hard to imagine anyone writing this), but it does expose the inner workings of the pass. Despite the lack of slices, even after the simplifying, there still seems to be a redundant statement or two. To understand why, let's remind ourselves what `RemoveUnused` actually does.

Conceptually, it's intended to skim through all the assignments in the program and remove those not found in the `HasUses::used` map. Using the terminology from before, this means "remove all the floaters who aren't anchored".
But what if the floater being removed was anchoring some earlier statement (previously a floater) ? That's exactly what happened in both examples.

C and D are saved thanks to `FindUninitialized::checkOutParameters`, so they get to stay (I won't explain how, it's not relevant to the point). B gets thrown out due to having no anchor later on (C doesn't do this, no `LocationSet` to read from), but **statement A stays because statement B was its anchor**. The `h.data` on the rhs of B makes sure of this anchoring. 

So, `RemoveUnused` does what it's told and in effect, kills off the only anchor (use) statement A was having. Thus A becomes an _orphaned_ floater - a useful statement that became redundant once we removed the (known) redundant statements.

I am not entirely sure, but there may be a workaround to this. The simplest solution is to run this pass twice, but this might not scale well (how many times is enough? maybe some recursive strategy might help). 

The core problem with this pass (I think) is that we can't detect all the redundant statements in one go. We need `RemoveUnused` to expose the orphaned floaters and then remove them or find a way to detect them before we reach the removing stage.

Reply:
SimplifyDefUse currently removes unused assignments, but it does not remove transitively unused assignments.
This can be fixed simply by converting the pass from a PassManager to a PassRepeated, and then it will be executed until a fixed-point is reached.